libbfddp/bfddp_extra.c

```diff
@@ -0,0 +1,858 @@
+
+/*
+ * Session functions.
+ */
+static void
+bfddp_session_set_slowstart(struct bfd_session *bs)
+{
+	/* Slow start settings. */
+	bs->bs_cur_dmultiplier = SLOWSTART_DMULT;
+	bs->bs_cur_tx = SLOWSTART_TX;
+	bs->bs_cur_rx = SLOWSTART_RX;
+	bs->bs_cur_erx = SLOWSTART_ERX;
+}
+
+static void
+bfddp_session_reset_remote(struct bfd_session *bs)
+{
+	/* Default remote settings. */
+	bs->bs_rdmultiplier = SLOWSTART_DMULT;
+	bs->bs_rtx = SLOWSTART_TX;
+	bs->bs_rrx = SLOWSTART_RX;
+	bs->bs_rerx = SLOWSTART_ERX;
+}
+
+void
+bfddp_session_update(struct bfd_session *bs, void *arg,
+		     const struct bfddp_session *bds)
+{
+	bool timers_changed = false;
+	struct in_addr *ia;
+	uint16_t port;
+	uint32_t min_rx, min_tx;
+	uint32_t flags = ntohl(bds->flags);
+
+	/*
+	 * Load flags.
+	 *
+	 * NOTE: normalize boolean values (e.g. `!!`) so packet build functions
+	 * can use it with shift (e.g. (multihop << X)).
+	 */
+	bs->bs_admin_shutdown = !!(flags & SESSION_SHUTDOWN);
+	bs->bs_multihop = !!(flags & SESSION_MULTIHOP);
+	bs->bs_passive = !!(flags & SESSION_PASSIVE);
+	bs->bs_demand = !!(flags & SESSION_DEMAND);
+	bs->bs_cbit = !!(flags & SESSION_CBIT);
+	bs->bs_echo = !!(flags & SESSION_ECHO);
+
+	if (bs->bs_multihop)
+		port = htons(BFD_MULTI_HOP_PORT);
+	else
+		port = htons(BFD_SINGLE_HOP_PORT);
+
+	bs->bs_ipv4 = !(bds->flags & SESSION_IPV6);
+	if (bs->bs_ipv4) {
+		ia = (struct in_addr *)&bds->src;
+		bs->bs_src.bs_src_sin.sin_family = AF_INET;
+		bs->bs_src.bs_src_sin.sin_addr.s_addr = ia->s_addr;
+
+		ia = (struct in_addr *)&bds->dst;
+		bs->bs_dst.bs_dst_sin.sin_family = AF_INET;
+		bs->bs_dst.bs_dst_sin.sin_addr.s_addr = ia->s_addr;
+		bs->bs_dst.bs_dst_sin.sin_port = port;
+	} else {
+		bs->bs_src.bs_src_sin6.sin6_family = AF_INET6;
+		memcpy(&bs->bs_src.bs_src_sin6.sin6_addr, &bds->src,
+		       sizeof(struct in6_addr));
+		if (IN6_IS_ADDR_LINKLOCAL(&bs->bs_src.bs_src_sin6.sin6_addr))
+			bs->bs_src.bs_src_sin6.sin6_scope_id = bds->ifindex;
+
+		bs->bs_dst.bs_dst_sin6.sin6_family = AF_INET6;
+		memcpy(&bs->bs_dst.bs_dst_sin6.sin6_addr, &bds->dst,
+		       sizeof(struct in6_addr));
+		bs->bs_dst.bs_dst_sin6.sin6_port = port;
+		if (IN6_IS_ADDR_LINKLOCAL(&bs->bs_dst.bs_dst_sin6.sin6_addr))
+			bs->bs_dst.bs_dst_sin6.sin6_scope_id = bds->ifindex;
+	}
+
+	/* Load timers. */
+	min_tx = ntohl(bds->min_tx);
+	min_rx = ntohl(bds->min_rx);
+	if (bs->bs_tx != min_tx || bs->bs_rx != min_rx)
+		timers_changed = true;
+
+	bs->bs_tx = min_tx;
+	bs->bs_rx = min_rx;
+	bs->bs_erx = ntohl(bds->min_echo_rx);
+	bs->bs_hold = ntohl(bds->hold_time);
+	bs->bs_dmultiplier = bds->detect_mult;
+
+	bs->bs_minttl = bds->ttl;
+	bs->bs_ifindex = ntohl(bds->ifindex);
+	if (bds->ifname[0])
+		snprintf(bs->bs_ifname, sizeof(bs->bs_ifname), "%s",
+			 bds->ifname);
+
+	bfddp_callbacks.bc_session_update(bs, arg);
+
+	/* Handle administrative shutdown. */
+	if (bs->bs_admin_shutdown) {
+		/* Send a control packet to tell we are shutting down. */
+		bs->bs_state = STATE_ADMINDOWN;
```

Comment:
prefix `STATE_ADMINDOWN`? S/t like `BFDD_` or alike

Reply:
idem for other enums / macros

Reply:
Sorry, I did not understand your comment... You are suggesting to add prefix to all enumerations already defined?

Reply:
yup

Reply:
I think this is a bit too late for this change... It would cause tons of code change.